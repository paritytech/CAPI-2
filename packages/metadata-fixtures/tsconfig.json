{
  "extends": "../../tsconfig.base",
  "include": ["src"],
  "compilerOptions": {
    "baseUrl": "src",
    "resolveJsonModule": true,
    "paths": {
      "@/*": ["*"]
    }
  }
}
