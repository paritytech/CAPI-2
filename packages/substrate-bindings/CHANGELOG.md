# Changelog

## Unreleased

### Breaking

- Move the `is` and `as` methods from enum values to `Enum`: `Enum.is(value, tag)`, `Enum.as(value, tag)`.

## 0.3.0 - 2024-04-25

### Added

- New type `EnumVariant<T, K>` to select one specific variant from an enum.
- Improved Enum type inference so that they can be assigned between types as long as they are compatible.

### Breaking

- The generic for `Enum<T>` is now an object of `{ [type: string]: any }`, rather than a union of `{ type: string, value: any } | ... | { type: string, value: any }`

## 0.2.0 - 2024-04-23

### Added

- Support for metadata V14

## 0.1.0 - 2024-04-22

### Changed

- Added support for `AccountId20`

## 0.0.1 - 2024-04-03

### Changed

Initial release
