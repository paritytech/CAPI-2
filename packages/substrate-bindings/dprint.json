{
  "incremental": true,
  "indentWidth": 2,
  "lineWidth": 80,
  "typescript": {
    "trailingCommas": "onlyMultiLine",
    "quoteProps": "asNeeded",
    "arrowFunction.useParentheses": "force",
    "semiColons": "asi"
  },
  "markdown": {
    "textWrap": "always"
  },
  "includes": ["**.{json,md,ts}"],
  "excludes": ["dist", "package-lock.json"],
  "plugins": [
    "https://plugins.dprint.dev/typescript-0.84.4.wasm",
    "https://plugins.dprint.dev/json-0.17.2.wasm",
    "https://plugins.dprint.dev/markdown-0.15.2.wasm"
  ]
}
