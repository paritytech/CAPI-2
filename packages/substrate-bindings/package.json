{
  "name": "@polkadot-api/substrate-bindings",
  "version": "0.6.0",
  "author": "Josep M Sobrepere (https://github.com/josepot)",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/polkadot-api/polkadot-api.git"
  },
  "license": "MIT",
  "sideEffects": false,
  "exports": {
    ".": {
      "node": {
        "production": {
          "import": "./dist/esm/index.mjs",
          "require": "./dist/min/index.js",
          "default": "./dist/index.js"
        },
        "import": "./dist/esm/index.mjs",
        "require": "./dist/index.js",
        "default": "./dist/index.js"
      },
      "module": "./dist/esm/index.mjs",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/index.js",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/index.js",
  "module": "./dist/esm/index.mjs",
  "browser": "./dist/esm/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc --noEmit && rollup -c ../../rollup.config.js",
    "test": "vitest",
    "lint": "prettier --check README.md \"src/**/*.{js,jsx,ts,tsx,json,md}\"",
    "format": "prettier --write README.md \"src/**/*.{js,jsx,ts,tsx,json,md}\"",
    "prepack": "pnpm run build"
  },
  "dependencies": {
    "@noble/hashes": "^1.3.1",
    "@scure/base": "^1.1.1",
    "@polkadot-api/utils": "workspace:*",
    "scale-ts": "^1.6.0"
  },
  "devDependencies": {
    "bigint-conversion": "^2.4.1",
    "hash-wasm": "^4.9.0"
  }
}
