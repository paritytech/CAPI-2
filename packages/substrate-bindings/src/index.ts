export * from "./codecs"
export * from "./hashes"
export * from "./storage"
export * from "./descriptors"
