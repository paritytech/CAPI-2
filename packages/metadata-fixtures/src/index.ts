import { V15, v15 } from "@polkadot-api/substrate-bindings"
import { readFile } from "node:fs/promises"
import { join } from "node:path"

export async function getKsmMetadataBuffer(baseFolder?: string) {
  const buffer = await readFile(
    join(baseFolder ?? __dirname, "../ksm-metadata.scale"),
  )
  return new Uint8Array(buffer)
}
export async function getKsmMetadata(baseFolder?: string): Promise<V15> {
  const buffer = await getKsmMetadataBuffer(baseFolder)
  return v15.dec(buffer)
}
