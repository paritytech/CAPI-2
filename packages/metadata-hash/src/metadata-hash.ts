import { Blake3256, V15, h64, v15 } from "@polkadot-api/substrate-bindings"
import { mergeUint8 } from "@polkadot-api/utils"
import {
  ExtraInfo,
  ExtrinsicMetadata,
  Lookup,
  MetadataDigest,
  ScalePrimitive,
  TypeDef,
  TypeRef,
  extraInfo,
  extrinsicMetadata,
  lookupType,
  metadataDigest,
} from "./codecs"

// h64(scale(metadata|tokenSymbol|tokenDecimals)): metadataDigest
const cache = new Map<bigint, Uint8Array>()

const pruneAndCollectAccessibleTypes = (
  metadata: V15,
  accessibleTypes: number[],
  id: number,
) => {
  if (accessibleTypes.indexOf(id) !== -1) return
  const { def } = metadata.lookup.find((x) => x.id === id)!
  switch (def.tag) {
    case "composite":
      if (def.value.length !== 0) {
        accessibleTypes.push(id)
        def.value.forEach(({ type }) =>
          pruneAndCollectAccessibleTypes(metadata, accessibleTypes, type),
        )
      }
      break
    case "variant":
      if (def.value.length !== 0) {
        accessibleTypes.push(id)
        def.value.forEach(({ fields }) =>
          fields.forEach(({ type }) =>
            pruneAndCollectAccessibleTypes(metadata, accessibleTypes, type),
          ),
        )
      }
      break
    case "sequence":
      accessibleTypes.push(id)
      pruneAndCollectAccessibleTypes(metadata, accessibleTypes, def.value)
      break
    case "array":
      accessibleTypes.push(id)
      pruneAndCollectAccessibleTypes(metadata, accessibleTypes, def.value.type)
      break
    case "tuple":
      if (def.value.length !== 0) {
        accessibleTypes.push(id)
        def.value.forEach((type) =>
          pruneAndCollectAccessibleTypes(metadata, accessibleTypes, type),
        )
      }
      break
    case "bitSequence": // bitSequence inner types are not stored
      accessibleTypes.push(id)
      break
    // primitive and compact are not stored
    case "primitive":
    case "compact":
      break
  }
}

const collectPrimitives = (
  metadata: V15,
  primitives: ScalePrimitive["tag"][],
  alreadyVisited: number[],
  frameId: number,
) => {
  if (alreadyVisited.includes(frameId)) return
  alreadyVisited.push(frameId)

  const { def } = metadata.lookup.find((x) => x.id === frameId)!
  switch (def.tag) {
    case "composite":
      def.value.forEach(({ type }) =>
        collectPrimitives(metadata, primitives, alreadyVisited, type),
      )
      break
    case "variant":
      def.value.forEach(({ fields }) =>
        fields.forEach(({ type }) =>
          collectPrimitives(metadata, primitives, alreadyVisited, type),
        ),
      )
      break
    case "array":
      break
    case "tuple":
      def.value.forEach((f) =>
        collectPrimitives(metadata, primitives, alreadyVisited, f),
      )
      break
    case "primitive":
      primitives.push(def.value.tag)
      break
    case "sequence":
    case "compact":
      collectPrimitives(metadata, primitives, alreadyVisited, def.value)
      break
    case "bitSequence":
      collectPrimitives(
        metadata,
        primitives,
        alreadyVisited,
        def.value.bitOrderType,
      )
      collectPrimitives(
        metadata,
        primitives,
        alreadyVisited,
        def.value.bitStoreType,
      )
      break
  }
}

const getTypeRef = (
  metadata: V15,
  accessibleTypes: number[],
  frameId: number,
): TypeRef => {
  const { def } = metadata.lookup.find((x) => x.id === frameId)!

  switch (def.tag) {
    case "composite":
    case "variant":
    case "tuple":
      if (accessibleTypes.includes(frameId))
        return { tag: "perId", value: accessibleTypes.indexOf(frameId) }
      else return { tag: "void", value: undefined }
    case "sequence":
    case "array":
    case "bitSequence":
      return { tag: "perId", value: accessibleTypes.indexOf(frameId) }
    case "primitive":
      // MerklePrimitive is a superset of ScalePrimitive
      return { tag: def.value.tag, value: undefined }
  }
  // from here on we have a compact
  const primitives: ScalePrimitive["tag"][] = []
  collectPrimitives(metadata, primitives, [], frameId)
  if (primitives.length > 1)
    throw new Error("Expected to find exactly 0 or 1 primitive")
  if (primitives.length === 0) return { tag: "void", value: undefined }
  switch (primitives[0]) {
    case "u8":
      return { tag: "compactU8", value: undefined }
    case "u16":
      return { tag: "compactU16", value: undefined }
    case "u32":
      return { tag: "compactU32", value: undefined }
    case "u64":
      return { tag: "compactU64", value: undefined }
    case "u128":
      return { tag: "compactU128", value: undefined }
    case "u256":
      return { tag: "compactU256", value: undefined }
    default:
      throw new Error("Invalid primitive for Compact")
  }
}

const constructTypeDef = (
  metadata: V15,
  accessibleTypes: number[],
  frameId: number,
): TypeDef[] => {
  const { def } = metadata.lookup.find((x) => x.id === frameId)!
  switch (def.tag) {
    case "composite":
      if (def.value.length === 0)
        throw new Error("Empty composites should be filtered out")
      return [
        {
          tag: "composite",
          value: def.value.map((f) => ({
            name: f.name,
            typeName: f.typeName,
            ty: getTypeRef(metadata, accessibleTypes, f.type),
          })),
        },
      ]
    case "variant": {
      if (def.value.length === 0)
        throw new Error("Empty composites should be filtered out")
      return def.value.map((v) => ({
        tag: "enumeration" as "enumeration",
        value: {
          name: v.name,
          index: v.index,
          fields: v.fields.map((f) => ({
            name: f.name,
            typeName: f.typeName,
            ty: getTypeRef(metadata, accessibleTypes, f.type),
          })),
        },
      }))
    }
    case "sequence":
      return [
        {
          tag: "sequence",
          value: getTypeRef(metadata, accessibleTypes, def.value),
        },
      ]
    case "array":
      return [
        {
          tag: "array",
          value: {
            len: def.value.len,
            typeParam: getTypeRef(metadata, accessibleTypes, def.value.type),
          },
        },
      ]
    case "tuple":
      if (def.value.length === 0)
        throw new Error("Empty tuples should be filtered out")
      return [
        {
          tag: "tuple",
          value: def.value.map((id) =>
            getTypeRef(metadata, accessibleTypes, id),
          ),
        },
      ]
    case "bitSequence": {
      const primitives: ScalePrimitive["tag"][] = []
      collectPrimitives(metadata, primitives, [], def.value.bitStoreType)
      if (primitives.length !== 1)
        throw new Error("Expected to find exactly 1 primitive")
      let numBytes
      switch (primitives[0]) {
        case "u8":
          numBytes = 1
          break
        case "u16":
          numBytes = 2
          break
        case "u32":
          numBytes = 4
          break
        case "u64":
          numBytes = 8
          break
        default:
          throw new Error("Invalid primitive for BitSequence")
      }
      const bitOrderType = def.value.bitOrderType
      const storeOrderPath = metadata.lookup.find(
        (x) => x.id === bitOrderType,
      )!.path
      let leastSignificantBitFirst
      if (storeOrderPath.includes("Lsb0")) leastSignificantBitFirst = true
      else if (storeOrderPath.includes("Msb0")) leastSignificantBitFirst = false
      else throw new Error("BitOrderType not recognized")
      return [
        {
          tag: "bitSequence",
          value: { numBytes, leastSignificantBitFirst },
        },
      ]
    }
    case "primitive":
    case "compact":
      throw new Error("Primitives and Compacts should be filtered out")
  }
}

export const buildMetadataHash = (
  metadata: V15,
  info: ExtraInfo,
): Uint8Array => {
  const keyHash = h64(mergeUint8(v15.enc(metadata), extraInfo.enc(info)))
  if (cache.has(keyHash)) {
    return cache.get(keyHash)!
  }

  const accessibleTypes: number[] = []
  pruneAndCollectAccessibleTypes(
    metadata,
    accessibleTypes,
    metadata.extrinsic.call,
  )
  pruneAndCollectAccessibleTypes(
    metadata,
    accessibleTypes,
    metadata.extrinsic.address,
  )
  pruneAndCollectAccessibleTypes(
    metadata,
    accessibleTypes,
    metadata.extrinsic.signature,
  )
  metadata.extrinsic.signedExtensions.forEach((se) => {
    pruneAndCollectAccessibleTypes(metadata, accessibleTypes, se.type)
    pruneAndCollectAccessibleTypes(
      metadata,
      accessibleTypes,
      se.additionalSigned,
    )
  })
  accessibleTypes.sort((a, b) => a - b)

  const typeTree: Lookup = []
  accessibleTypes.forEach((ty, idx) => {
    const frameType = metadata.lookup.find((x) => x.id === ty)!
    constructTypeDef(metadata, accessibleTypes, ty).forEach((tyDef) => {
      typeTree.push({
        path: frameType.path,
        typeId: idx,
        typeDef: tyDef,
      })
    })
  })

  typeTree.sort((a, b) => {
    if (a.typeId !== b.typeId) return a.typeId - b.typeId // in general
    // should only happen for variants
    if (a.typeDef.tag !== "enumeration" || b.typeDef.tag !== "enumeration")
      throw new Error("Found two types with same id")
    return a.typeDef.value.index - b.typeDef.value.index
  })

  // let's build the lookup hash
  const nodes = typeTree.map((x) => Blake3256(lookupType.enc(x)))
  while (nodes.length > 1) {
    const right = nodes.pop()!
    const left = nodes.pop()!
    nodes.unshift(Blake3256(mergeUint8(left, right)))
  }
  // if no types in the tree, we need to return 0's array
  const rootLookupHash =
    nodes.length === 1 ? nodes.pop()! : new Uint8Array(32).fill(0)
  const extrinsic: ExtrinsicMetadata = {
    version: metadata.extrinsic.version,
    addressTy: getTypeRef(
      metadata,
      accessibleTypes,
      metadata.extrinsic.address,
    ),
    callTy: getTypeRef(metadata, accessibleTypes, metadata.extrinsic.call),
    signatureTy: getTypeRef(
      metadata,
      accessibleTypes,
      metadata.extrinsic.signature,
    ),
    signedExtensions: metadata.extrinsic.signedExtensions.map((se) => ({
      identifier: se.identifier,
      includedInExtrinsic: getTypeRef(metadata, accessibleTypes, se.type),
      includedInSignedData: getTypeRef(
        metadata,
        accessibleTypes,
        se.additionalSigned,
      ),
    })),
  }

  const digest: MetadataDigest = {
    tag: "V1",
    value: {
      typeInformationTreeRoot: rootLookupHash,
      extrinsicMetadataHash: Blake3256(extrinsicMetadata.enc(extrinsic)),
      ...info,
    },
  }
  return Blake3256(metadataDigest.enc(digest))
}
