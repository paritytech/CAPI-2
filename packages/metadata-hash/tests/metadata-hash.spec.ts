import { buildMetadataHash } from "@/metadata-hash"
import { getKsmMetadata } from "@polkadot-api/metadata-fixtures"
import { metadata } from "@polkadot-api/substrate-bindings"
import { beforeAll, describe, expect, it } from "vitest"

let ksmMetadata: Awaited<Uint8Array>
beforeAll(async () => {
  ksmMetadata = metadata.enc({
    magicNumber: 0x6174656d,
    metadata: {
      tag: "v15",
      value: await getKsmMetadata(),
    },
  })
})

describe("getChecksumBuilder snapshots", () => {
  it("generates metadata hash", () => {
    const result = buildMetadataHash(ksmMetadata, {
      decimals: 1,
      specName: "nice",
      base58Prefix: 1,
      specVersion: 1,
      tokenSymbol: "lol",
    })
    expect(result).toMatchSnapshot()
  })
})
