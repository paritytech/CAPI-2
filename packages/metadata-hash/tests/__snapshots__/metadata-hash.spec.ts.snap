// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getChecksumBuilder snapshots > generates metadata hash 1`] = `
Uint8Array [
  184,
  178,
  23,
  143,
  51,
  255,
  30,
  57,
  173,
  177,
  0,
  244,
  198,
  0,
  95,
  218,
  107,
  66,
  94,
  93,
  216,
  143,
  90,
  110,
  194,
  169,
  216,
  40,
  248,
  95,
  98,
  198,
]
`;
